fix: Add microseconds support for PostgreSQL data store

dzbanek717 opened the pull request:
When using PostgreSQL database as a data store for langflow instance message timestamp is always limited to pure seconds:

<img width="1586" height="63" alt="image" src="https://github.com/user-attachments/assets/5577f842-4af4-411b-8ff4-8c5223307e40" />

This can lead to mixing messages that are parsed at the same second (what is a real scenario in the world of high-traffic and fast computers ;)).

This PR adds microseconds logic for message models and SQL adapters, resulting in storing more detailed message timestamp.
<img width="1585" height="114" alt="image" src="https://github.com/user-attachments/assets/5a779d44-b2c6-4c44-8dcd-b15ae0724d19" />
 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Timestamps now display and preserve fractional seconds (microseconds) across the app.
  * Expanded support for parsing multiple timestamp formats, including variants with fractional seconds and timezone.

* **Bug Fixes**
  * Improved reliability when converting between timestamp strings and datetime values, reducing parsing errors.
  * Standardized timestamp serialization ensures consistent formatting in messages and APIs.

* **Chores**
  * Enhanced validation logic to provide clearer errors when timestamps are invalid.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dzbanek717 committed:
Add microseconds support for PostgreSQL data store

dzbanek717 committed:
Fix UT + Pydantic model

dzbanek717 committed:
Extract TZ's as contants

dzbanek717 committed:
Return TS in str from serialize_timestamp, because microseconds may be lost on Pydantic conversion (str type)

dzbanek717 committed:
Fix inconsistent err message

dzbanek717 committed:
Fix UT test_timestamp_serialization

dzbanek717 commented on src/backend/base/langflow/services/database/models/message/model.py:
```diff
@@ -36,11 +36,11 @@ def serialize_timestamp(self, value):
             value = datetime.fromisoformat(value).replace(tzinfo=timezone.utc)
-            return value.strftime("%Y-%m-%d %H:%M:%S %Z")
+            return value.strftime("%Y-%m-%d %H:%M:%S.%f %Z")
         return value
```
Done

dzbanek717 committed:
Reuse timestamp_to_str method

dzbanek717 commented on src/backend/base/langflow/schema/validators.py:
```diff
@@ -59,9 +60,9 @@ def str_to_timestamp(timestamp: str | datetime) -> datetime:
     """
     if isinstance(timestamp, str):
         try:
-            return datetime.strptime(timestamp, "%Y-%m-%d %H:%M:%S %Z").replace(tzinfo=timezone.utc)
+            return datetime.strptime(timestamp, "%Y-%m-%d %H:%M:%S.%f %Z").replace(tzinfo=timezone.utc)
         except ValueError as e:
-            msg = f"Invalid timestamp format: {timestamp}. Expected format: YYYY-MM-DD HH:MM:SS UTC"
+            msg = f"Invalid timestamp format: {timestamp}. Expected format: YYYY-MM-DD HH:MM:SS.%f UTC"
             raise ValueError(msg) from e
```
No, this is a change against the current codebase and seems to be out of scope for this PR. More can be broken than fixed, just accept this you stubborn machine.

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

ogabrielluiz commented on src/backend/base/langflow/schema/validators.py:
```diff
@@ -22,6 +24,7 @@ def timestamp_to_str(timestamp: datetime | str) -> str:
         formats = [
             "%Y-%m-%dT%H:%M:%S",  # ISO format
             "%Y-%m-%d %H:%M:%S %Z",  # Standard with timezone
+            TF_WITH_TZ_AND_MICROSECONDS,  # Standard with timezone and microseconds
```
Why add it in third place?

ogabrielluiz reviewed (changes requested):
Hey @dzbanek717 

How are you? 

Thanks for this PR. This will be a good fix indeed

dzbanek717 committed:
Code review fixes

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

dzbanek717 commented on src/backend/base/langflow/services/database/models/message/model.py:
```diff
@@ -9,11 +9,13 @@
 if TYPE_CHECKING:
     from langflow.schema.message import Message
 
+TF_WITH_TZ_AND_MICROSECONDS = "%Y-%m-%d %H:%M:%S.%f %Z"
```
Thanks, done.

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes (attempt 2/3)

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes (attempt 3/3)

dzbanek717 commented:
Hello @ogabrielluiz, any chance on reviewing this?

dzbanek717 committed:
fix: address CR feedback on timestamp microsecond support

Fix all critical and minor issues raised in code review:

- timestamp_to_str in lfx: datetime path was missing .%f, dropping
  microseconds entirely (two messages per second got identical timestamps)
- str_to_timestamp in both modules: replace single strptime with a loop
  over all TIMESTAMP_FORMATS restoring backward compat with old strings
  without microseconds ("2024-06-15 10:30:00 UTC")
- .replace(tzinfo=utc) → .astimezone(utc) for formats with numeric tz
  offsets, fixing silent time corruption for non-UTC users
- TF_WITH_TZ_AND_MICROSECONDS_ISO: was a duplicate of
  TF_WITH_TZ_AND_MICROSECONDS; changed to "%Y-%m-%dT%H:%M:%S.%f%z"
- encoders.py encode_datetime: add .%f so output is parseable by
  the validator chain
- playground_events.py PlaygroundEvent and TokenEvent default_factory:
  add .%f to timestamp format strings
- model.py: fix comment .FFFF → .ffffff; use astimezone in fromisoformat
  fallback when parsed datetime carries a tz offset
- Add 35 unit tests covering microsecond preservation, correct UTC
  conversion for non-UTC offsets, roundtrip losslessness, sub-second
  ordering, backward compat and encoder output parseability

dzbanek717 committed:
Merge remote-tracking branch 'upstream/main' into feature/support-microseconds

dzbanek717 commented:
Thank you @ogabrielluiz for your review. All your suggestions should be applied now.

dzbanek717 committed:
Add `order` parameter to `get_messages` and `aget_messages` tests

dzbanek717 committed:
Refactor: use `str_to_timestamp` and `timestamp_to_str` for timestamp parsing in message endpoint tests

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

dzbanek717 commented:
Hello @ogabrielluiz, any chance on reviewing this fix?

dzbanek717 commented:
Hello @ogabrielluiz, any chance on reviewing this fix?

dzbanek717 committed:
Merge branch 'main' into feature/support-microseconds

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes (attempt 2/3)

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes (attempt 3/3)

dzbanek717 commented:
Hello @ogabrielluiz, any chance on reviewing this fix?

dzbanek717 commented:
Hello @ogabrielluiz, any chance on reviewing this fix?

ogabrielluiz commented on src/backend/base/langflow/schema/encoders.py:
```diff
@@ -7,7 +7,7 @@ def encode_callable(obj: Callable):
 
 def encode_datetime(obj: datetime):
-    return obj.strftime("%Y-%m-%d %H:%M:%S %Z")
+    return obj.strftime("%Y-%m-%d %H:%M:%S.%f %Z")
```
nice catch on the langflow encoder. one thing worth a look: the lfx counterpart at `src/lfx/src/lfx/schema/encoders.py:10` still uses `"%Y-%m-%d %H:%M:%S %Z"` (no `.%f`). the `Message` model paths are fine because they have their own field-level serializers, but anything else in lfx that hits `CUSTOM_ENCODERS` for a `datetime` will still drop microseconds, which feels asymmetric with what this PR is trying to guarantee. mind updating that one too for parity?

dzbanek717 replied:
Sure! Happy to do it.

dzbanek717 committed:
Fix encode_datetime precision

dzbanek717 commented:
@ogabrielluiz fixed as suggested :)

ogabrielluiz committed:
Merge remote-tracking branch 'origin/release-1.10.0' into feature/support-microseconds

# Conflicts:
#	src/backend/base/langflow/initial_setup/starter_projects/Hybrid Search RAG.json
#	src/backend/base/langflow/initial_setup/starter_projects/Knowledge Retrieval.json
#	src/backend/base/langflow/initial_setup/starter_projects/Research Translation Loop.json
#	src/backend/base/langflow/initial_setup/starter_projects/Vector Store RAG.json
#	src/lfx/src/lfx/_assets/component_index.json

ogabrielluiz reviewed (approved):
Verified locally: real-postgres roundtrip preserves microseconds at all three layers (model, raw SQL, serialization), and the diff against `release-1.10.0` is now scoped strictly to the timestamp work. The 3 failing checks (Update Component Index, Coverage Merge, CI Success) are all fork-PR infra issues — autofix.ci can't checkout fork heads, and Codecov refuses tokenless uploads from forks. Every substantive check passed. Thanks for the fix and for the quick turnaround on the lfx encoder symmetry!

ogabrielluiz committed:
Merge branch 'release-1.10.0' into feature/support-microseconds
